tests/infer/mcmc/test_hmc.py

```diff
@@ -234,3 +234,69 @@ def model(data):
         posterior.append(trace.nodes['p_latent']['value'])
     posterior_mean = torch.mean(torch.stack(posterior), 0)
     assert_equal(posterior_mean, true_probs, prec=0.02)
+
+
+def test_logistic_regression_with_dual_averaging():
+    dim = 3
+    true_coefs = Variable(torch.arange(1, dim+1))
+    data = Variable(torch.randn(2000, dim))
+    labels = dist.Bernoulli(logits=(true_coefs * data).sum(-1)).sample()
+
+    def model(data):
+        coefs_mean = Variable(torch.zeros(dim), requires_grad=True)
+        coefs = pyro.sample('beta', dist.Normal(coefs_mean, Variable(torch.ones(dim))))
+        y = pyro.sample('y', dist.Bernoulli(logits=(coefs * data).sum(-1)), obs=labels)
+        return y
+
+    hmc_kernel = HMC(model, step_size=None)
+    hmc_kernel._adapted = True
```

Comment:
I think a nice API to have for HMC would be as follows:
 - `HMC(model)`: uses NUTS kernel with dual averaging.
 - `HMC(model, step_size=x)`: uses NUTS kernel without dual averaging.
 - `HMC(model, num_steps=y)`: uses HMC kernel with dual averaging.
 - `HMC(model, step_size=x, num_steps=y)`: uses vanilla HMC kernel.

With this API, we should also think about the default `warmup_steps` for the cases that we are estimating the `step_size` (it shouldn't be 0). However, we can handle this in a separate PR that cleans out the API. For the time being, we should definitely remove the `_adapted` and `_trajectory_length` attributes, and pass them into the constructor if needed.